A Hub API that can query all running sessions.

krmahadevan opened the pull request:
Closes #6070

The API can be accessed via 
http://<IP:Port>/grid/api/sessions.

Both GET and POST are supported and result in the
same behavior.

A typical response can look like below (intentionally
trimmed the capabilities to reduce verbosity)

```json
{
  "proxies": [
    {
      "proxyId": "http://192.168.1.6:5555",
      "proxyRemoteHost": "http://192.168.1.6:5555",
      "sessions": {
        "status": 0,
        "value": [
          {
            "id": "df8aa16d-1d52-2e43-b584-2f9df586b0f3",
            "capabilities": {}
          },
          {
            "id": "f85f5641-e6de-3642-913a-a45aafdb85ad",
            "capabilities": {}
          }
        ]
      }
    },
    {
      "proxyId": "http://192.168.1.6:5556",
      "proxyRemoteHost": "http://192.168.1.6:5556",
      "sessions": {
        "status": 0,
        "value": [
          {
            "id": "f4df4fbd-96ee-e042-ab43-40a33698d1e9",
            "capabilities": {}
          }
        ]
      }
    }
  ],
  "success": true
}
```

- [X] By placing an `X` in the preceding checkbox, I verify that I have signed the [Contributor License Agreement](https://github.com/SeleniumHQ/selenium/blob/master/CONTRIBUTING.md#step-6-sign-the-cla)

krmahadevan commented:
ping @shs96c @barancev - Can one of you please help take a look at this ?

shs96c commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+        continue;
+      }
+      Map<String, Object> res = new TreeMap<>();
+      res.put("proxyId", proxy.getId());
```
This is already going to be returned as a list of proxies. Having the word "`proxy`" seems redundant.

krmahadevan replied:
The current sessions payload, has no information about the proxy. That is why added both the proxyId and proxyRemoteHost so that the user gets to know as to for which proxy does the sessions info belong to.

shs96c commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+      try (InputStream in = new ByteArrayInputStream(asBytes(rsp));
+           Reader reader = new InputStreamReader(in, getContentEncoding(rsp))) {
+        Object body = new Json().newInput(reader).read(Object.class);
+        if (body instanceof Map) {
```
Instead of doing this, you can get the json coercer to do all the hard work:

```
Map<String, Object> body = json.newInput(reader).read(Json.MAP_TYPE);
```

Also, you're going to leak the `JsonInput`: best open that in the try statement above.

shs96c commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+    } catch (IOException e) {
+      throw new GridException(e.getMessage());
+    }
+    return new TreeMap<>();
```
We don't want to indicate some form of error has occurred?

krmahadevan replied:
Fixed it. The status gets flipped to `false` if there was any difficulty in extracting sessions info

shs96c commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  private static byte[] asBytes(HttpResponse rsp) {
+    InputStream stream = rsp.consumeContentStream();
```
Heavens. Just call `rsp.getContent()`. You're not saving any memory doing this.

krmahadevan replied:
Fixed. I think I borrowed that code from somewhere :)

shs96c commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+    }
+  }
+
+  private static Charset getContentEncoding(HttpResponse rsp) {
```
And this is `rsp.getContentEncoding()`

krmahadevan replied:
fixed.

krmahadevan commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+    if (req.getInputStream() != null) {
+      process(resp);
+    } else {
+      process(resp);
```
Fixed it. Initially I was thinking of supporting something like giving the ability to query sessions given a node. Forgot to remove the branching, when I didn't go forward with the implementation.

krmahadevan commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+    response.setCharacterEncoding("UTF-8");
+    response.setStatus(200);
+    Map<String, Object> proxies = new TreeMap<>();
+    try (Writer writer = response.getWriter(); JsonOutput out = json.newOutput(writer)) {
```
Fixed.

krmahadevan commented on java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java:
```diff
@@ -0,0 +1,154 @@
+      HttpResponse rsp = proxy.getHttpClient(url).execute(req);
+      try (InputStream in = new ByteArrayInputStream(asBytes(rsp));
+           Reader reader = new InputStreamReader(in, getContentEncoding(rsp))) {
+        Object body = new Json().newInput(reader).read(Object.class);
```
Fixed.

krmahadevan commented:
@shs96c 

> This is missing tests, so it'll be easy to introduce accidental regressions.

I was initially confused as to what sort of tests I add because this new servlet internally makes a call to `WebDriverServlet` running in the node. So didn't know how to fake that out. Ended up writing a E2E test which tests the functionality by spinning a grid hub and node.

> The approach taken here is very "bare metal": we're basically exposing the internals of the current grid to the world. Is this the approach we want to take, or would we prefer something that explains the domain better? I'm perfectly happy if the answer is "bare metal to get us off the ground", as I think we're allowed to iterate on Grid functionality with greater freedom than anything on the client side.

The ask in #6070  was explicitly to have the capability of listing out all the sessions running in the Hub via an API. Please let me know if there's any other way of giving this capability. I will alter my PR accordingly.

I have addressed the review comments. Please help take a look.

krmahadevan commented:
@shs96c - thank you so much for merging the other two PRs. Can you please check if this one needs any additional work or if it can be merged as well ?

krmahadevan commented:
Rebased off of `master` and pushed in the updated branch.
